### DISCUSSION TITLE
This is awesome!

### DISCUSSION THREAD

**Discussion by kelvin124124** [sappho192/ffxiv-ja-ko-translator#14] [General]
Awesome work!
May I ask how much vram do I need to run this since I'm not familiar with LLM.
And do you have any plan for models on other in-game languages like En, Fr, De and Cn?

**Comment by sappho192** [comment]
Hi @kelvin124124 ! Thank you for contacting me.

In current state, roughly 3.2GB would be required to run the model. But it would be change after conducting some model optimizations.
Note that current example project [[GuiExample](https://github.com/sappho192/ffxiv-ja-ko-translator/blob/main/onnx_project/dotnet/GuiExample/FF14Translator.cs)] uses the CPU version of OnnxRumtime, you'll need to change the dependency to OnnxRumtime.GPU[[Link](https://www.nuget.org/packages/Microsoft.ML.OnnxRuntime.gpu)] if you want to run it on GPU.

Currently I don't have a plan for making other language-paired model. But if you want, we can have a discussion about it!

Do you have any translation pair in mind? (e.a. Ja→En)

**Reply by kelvin124124** [reply]
Good to know the model can run on CPU! Your project is perfectly in line with what an ideal in-game translator would be in my mind. As for other language-paired models, I do wish to see a Ja→En model one day if that doesn't bring much trouble to you because those are the two languages used by the main player base.

Btw maybe a model that extracts in-game vocabularies, get their translations and ship the translations to other LLMs like Llama or GPT-4 would be a better approach to creating the perfect translator? But I don't think that belongs to the category of LLM.

**Reply by sappho192** [reply]
> Good to know the model can run on CPU! Your project is perfectly in line with what an ideal in-game translator would be in my mind. 
Happy to hear that :) Advances in LLM ans NLP in nowadays (just for few years) made this objective to be possible. 

>As for other language-paired models, I do wish to see a Ja→En model one day if that doesn't bring much trouble to you because those are the two languages used by the main player base.

Ja→En would be easier to construct as there are bunch of accessible datasets and models. I'll try that in near future ' -')b

> Btw maybe a model that extracts in-game vocabularies, get their translations and ship the translations to other LLMs like Llama or GPT-4 would be a better approach to creating the perfect translator? But I don't think that belongs to the category of LLM.

That can be still regarded as LLM, but I would prefer fine-tuning the LLM models. But you need to know that current major small LLMs like 7B model takes 10GB of disk, 1.3B would take 4GB, which I think is a bit huge.

I've looked your projects(like ChatTranslated). If you are using ChatGPT pro plan, maybe you can make fine-tuned model based on ChatGPT 3.5 I think...? (But there's a downside that we need to pay OpenAI to use the fine-tuned model)
You can find Auto-translate texts within the game, ffxiv wiki, or github repos, which can be used as fine-tuning datasets.
(Of course there are slangs like エキルレ(Roullete: Experts), etc. Which should be gathered manually)

As I mentioned in README, current translation model is still in proof of concept phase, and fine-tuning is not properly done yet.
(Which means that current translation performace is not specialized in FF14)  
If we gather some datasets which contains FFXIV-specific words or expressions enough, this may be achieved easily.
Maybe I could make a website which receives FFXIV-related chat or sentence from anyone to make the dataset!

**Reply by kelvin124124** [reply]
>If we gather some datasets which contains FFXIV-specific words or expressions enough, this may be achieved easily.
>Maybe I could make a website which receives FFXIV-related chat or sentence from anyone to make the dataset!

I will consider collecting users' chat messages at the cost of providing a free GPT translation option to them, or just collect the messages by default and give them an option to opt-out but that sounds like spyware and is subjected to PAC's approval. 
Regarding the website that receives FFXIV-related chat, I think a serverless function to proxy the API requests to OpenAI (if we choose the first approach) and store the message in an online database is enough?

**Reply by sappho192** [reply]
> I will consider collecting users' chat messages at the cost of providing a free GPT translation option to them, or just collect the messages by default and give them an option to opt-out but that sounds like spyware and is subjected to PAC's approval. 

It would be a great help if you do so...! 
Since I also serve and maintain a [[translator app](https://github.com/sappho192/IronworksTranslator)] for FFXIV players, I've once thought the proper way to collect data from the users:

- Make users to choose one option from below:
  1) Collect and send all the chat data
  2) (default option?) Send the only chat data that the user choosed to
  3) Never collect or send the data 

For the option 2), I thought attatching a small send(or question mark?) button at the end of each chatlogs.

> Regarding the website that receives FFXIV-related chat, I think a serverless function to proxy the API requests to OpenAI (if we choose the first approach) and store the message in an online database is enough?

For that part, I've been already been trying to make a server stack to receive, store, and translate the FFXIV-related sentences from the user. 

Of course this is just an excerpt, but rough structure is like below:
![프레젠테이션1](https://github.com/sappho192/ffxiv-ja-ko-translator/assets/7788738/ad55bd65-9d3e-42a4-9f5c-301bd63c423e)
1. Your translator app could send the chat data to the DB server via REST API with JSON format. [[Candidate DB](https://pocketbase.io/)]
2. I (or we) periodically inspect the pending chat data (to filter out sensitive, inappropriate things) and add them to Weblate server.
![image](https://github.com/sappho192/ffxiv-ja-ko-translator/assets/7788738/7e05af27-c4ed-4535-98ca-c548d41077a3)
3. In Weblate server, the chat data can be translated into various language(JEDF+Ko) [[Demo](https://ff14.sapphosound.com/)]
4. Translated data will be backed up in Github repository. [[Demo](https://github.com/Akadaemia-Anyder/ff14-idioms/blob/main/src/ko.xliff)]

For step 1, I'm also considering to make a simple website to receive FFXIV-related sentences from anyone like below:
![image](https://github.com/sappho192/ffxiv-ja-ko-translator/assets/7788738/ee7ad86b-1100-4c05-8a10-9d72625647ef)

So far, it doesn't seem that hard to implement those system. Would this sound good for you too?

**Reply by kelvin124124** [reply]
Sounds good. Can I have an admin and an app account on the dB server to mess around and configure my serverless proxy function?

**Reply by kelvin124124** [reply]
You can drop me an email at kelvin@kelpcc.com

**Reply by sappho192** [reply]
Sure! 🤗  I will build a DB server and create an admin account around Sunday. (It's the Lunar New Year holiday season here.)
Thank you ' v')b

**Reply by sappho192** [reply]
@kelvin124124 I've setup the DB server and sent an email about your admin account and endpoint info. 
Tell me if you don't receive the email.

**Comment by sappho192** [comment]
@kelvin124124 

There are two choices for what the right model is for supporting multiple languages.
The most popular multilingual translation models(LLM-based) are between 4GB~10GB in size, which can be a bit overwhelming to use on a personal computer.

Instead, you can keep the model size to 1-2 GB by creating a single combination of translation models like the one I've shown here, and then only using the ones you need. (For example, downloading only En→Ja and Ja→En models in the program settings UI)

However, you will have to train the model for as many translation pairs as you want to support, so the initial cost will be high.
(For reference, the translation model released here was trained on a 48GB Nvidia RTX A6000, which took about 45GB of VRAM and 4 days to train, and the GPU was borrowed temporarily🥲).

**Comment by sappho192** [comment]
> May I ask how much vram do I need to run this since I'm not familiar with LLM. 

Hi @kelvin124124 , I have a good news for you. 
Today I tried optimizing my translation model using [[Quantization](https://huggingface.co/docs/optimum/onnxruntime/usage_guides/quantization#quantize-seq2seq-models)], and the result is pretty good.
The total model file size has reduced from 1.13GB to 0.30GB, and required RAM space also have shrinked from 3GB to 960MB.
Translation result is nearly same (but sometimes fails), so this quantized model can be provided as an 2nd option for low-spec PC users.

Around the end of February, I think I will train a baseline model of Ja→En translation. After that, fine-tuning using FFXIV-related texts will be proceeded. 🤞
